### PR TITLE
Outbound sequence number overriding

### PR DESCRIPTION
**Pull request by logist322**
Move outbound sequence number overriding on the transport level to prevent `SRTP` crashing on the `RTCRtpSender::replace_track()` call.

### CODE REVIEW

**Comment by melekes** [comment]
Is there an open issue? If not, @logist322 could you describe the problem in detail? Under what conditions does `SRTP` crash? How we can replicate?

**Comment by k0nserv** [comment]
@melekes more details are available here https://github.com/webrtc-rs/webrtc/issues/317

@logist322 I think this is a good start, but we need to address this concern

> For TrackLocalStaticRTP this might mean that the underlying transport layer overwrites the sequence number of packets passed to write_rtp or perhaps we have a way to signal in TrackLocalStaticRTP whether sequence numbers should be respected or not

From #317.

In particular, the code as written here only works when the RTP packets are guaranteed to be sent in order, an RTP forwarder would break with this code since it might forward packets out of order.

**Comment by logist322** [comment]
@k0nserv I guess it's ready for review :)

File: webrtc/src/rtp_transceiver/rtp_sender/mod.rs [hunk 1]
```diff
@@ -479,6 +487,18 @@ impl RTCRtpSender {
         self.internal.read_rtcp(self.receive_mtu).await
     }
 
+    /// Enables overriding outcoming `RTP` packets' `sequence number`s.
```
Comment (melekes) [inline webrtc/src/rtp_transceiver/rtp_sender/mod.rs:490]:
```suggestion
    /// Enables overriding outgoing `RTP` packets' `sequence number`s.
```

File: webrtc/src/rtp_transceiver/srtp_writer_future.rs [hunk 2]
```diff
@@ -12,6 +12,70 @@ use interceptor::{Attributes, RTCPReader, RTPWriter};
 use std::sync::atomic::{AtomicBool, Ordering};
 use std::sync::{Arc, Weak};
 use tokio::sync::Mutex;
+use util;
+
+/// RTP packet sequence number manager.
+pub(crate) struct SequenceTransformer(util::sync::Mutex<SequenceTransformerInner>);
+
+struct SequenceTransformerInner {
+    offset: u16,
```
Comment (melekes) [inline webrtc/src/rtp_transceiver/srtp_writer_future.rs:21]: would be good to add some comments here

File: webrtc/src/rtp_transceiver/srtp_writer_future.rs [hunk 3]
```diff
@@ -12,6 +12,70 @@ use interceptor::{Attributes, RTCPReader, RTPWriter};
...
+            return None;
+        }
+
+        let offset = guard
+            .reset_needed
+            .then(|| {
+                guard.reset_needed = false;
+                let offset = guard.last_sq.overflowing_sub(raw_sn.overflowing_sub(1).0).0;
+                guard.offset = offset;
+                offset
```
Comment (melekes) [inline webrtc/src/rtp_transceiver/srtp_writer_future.rs:70]: we don't need to reset guard.last_sq here, right?

File: webrtc/src/rtp_transceiver/rtp_sender/mod.rs (hunk 1 above)
Comment (logist322) [reply webrtc/src/rtp_transceiver/rtp_sender/mod.rs:490]: Oh, you are right

File: webrtc/src/rtp_transceiver/srtp_writer_future.rs (hunk 2 above)
Comment (logist322) [reply webrtc/src/rtp_transceiver/srtp_writer_future.rs:21]: I'll do it

File: webrtc/src/rtp_transceiver/srtp_writer_future.rs (hunk 3 above)
Comment (logist322) [reply webrtc/src/rtp_transceiver/srtp_writer_future.rs:70]: Right, here we just getting or calculating `offset`.

File: webrtc/src/rtp_transceiver/srtp_writer_future.rs (hunk 2 above)
Comment (logist322) [reply webrtc/src/rtp_transceiver/srtp_writer_future.rs:26]: @melekes I have added some docs

**Review by melekes — APPROVED** [review]
👍

**Comment by logist322** [comment]
@melekes as I can understand, you have approved this PR. So, can you merge it? :sweat_smile:

**Comment by logist322** [comment]
@melekes @k0nserv anyone, pls, merge it :sweat_smile:

**Comment by k0nserv** [comment]
Sorry @logist322 I've been super busy recently. Can you rebase on main so clippy and formatting corrects itself?

**Comment by logist322** [comment]
@k0nserv done!

File: webrtc/src/rtp_transceiver/srtp_writer_future.rs [hunk 4]
```diff
@@ -12,6 +12,92 @@ use interceptor::{Attributes, RTCPReader, RTPWriter};
...
+        self.0.lock().reset_needed = true;
+    }
+
+    /// Gets [`Some`] consistent `sequence number` if this [`SequenceTransformer`] is
+    /// enabled or [`None`] if it is not.
+    ///
+    /// Once this method is called, considers data sending began.
+    fn seq_number(&self, raw_sn: u16) -> Option<u16> {
+        let mut guard = self.0.lock();
+        guard.data_sent = true;
```
Comment (k0nserv) [inline webrtc/src/rtp_transceiver/srtp_writer_future.rs:80]: Shouldn't this be after `guard.enabled` below?

Comment (logist322) [reply webrtc/src/rtp_transceiver/srtp_writer_future.rs:80]: No, it shouldn't.
We have already discussed it with you in discord. If data is sent we are not allowed to enable `SequenceTransformer`.

Comment (logist322) [reply webrtc/src/rtp_transceiver/srtp_writer_future.rs:80]: So, we should register data sending even if `SequenceTransformer` is not `enabled`.

Comment (k0nserv) [reply webrtc/src/rtp_transceiver/srtp_writer_future.rs:80]: Ah, right I recall

**Pull request merged**